src/throttle/amoc_throttle_controller.erl

```diff
@@ -203,123 +205,131 @@ handle_info({change_plan, Name}, State) ->
-    CurrentRatePerMin = rate_per_minute(Info#throttle_info.rate, Info#throttle_info.interval),
-    ReqRatePerMin = rate_per_minute(Rate, Interval),
-    case {CurrentRatePerMin, Info#throttle_info.change_plan} of
-        {ReqRatePerMin, _} -> ok;
-        {_, undefined} -> do_change_rate(Name, Rate, Interval);
-        _ -> {error, cannot_change_rate}
-    end.
-
--spec do_change_rate(name(), amoc_throttle:rate(), amoc_throttle:interval()) -> ok | {error, any()}.
-do_change_rate(Name, Rate, Interval) ->
-    case amoc_throttle_process:get_throttle_processes(Name) of
-        {ok, List} ->
-            RatePerMinute = rate_per_minute(Rate, Interval),
-            report_rate(Name, RatePerMinute),
-            update_throttle_processes(List, Interval, Rate, length(List)),
-            ok;
-        Error ->
-            Error
-    end.
-
--spec start_gradual_rate_change(
-        name(), amoc_throttle:rate(), amoc_throttle:rate(),
-        amoc_throttle:interval(), pos_integer(), pos_integer(), throttle_info()) ->
-    throttle_info().
-start_gradual_rate_change(Name, LowRate, HighRate, RateInterval, StepInterval, NoOfSteps, Info) ->
-    ok = do_change_rate(Name, LowRate, RateInterval),
-    {ok, Timer} = timer:send_interval(StepInterval, {change_plan, Name}),
-    Plan = #change_rate_plan{high_rate = HighRate, no_of_steps = NoOfSteps, timer = Timer},
-    Info#throttle_info{rate = LowRate, interval = RateInterval, change_plan = Plan}.
-
-start_throttle_processes(Name, Interval, Rate, N) ->
-    ok = amoc_throttle_pool:start_process_pool(Name, Interval, Rate, N).
-
-update_throttle_processes([Pid], Interval, Rate, 1) ->
-    amoc_throttle_process:update(Pid, Interval, Rate);
-update_throttle_processes([Pid | Tail], Interval, Rate, N) when N > 1 ->
-    ProcessRate = Rate div N,
-    amoc_throttle_process:update(Pid, Interval, ProcessRate),
-    update_throttle_processes(Tail, Interval, Rate - ProcessRate, N - 1).
-
-run_in_all_processes(Name, Cmd) ->
-    case amoc_throttle_process:get_throttle_processes(Name) of
-        {ok, List} ->
-            [run_cmd(P, Cmd) || P <- List],
-            ok;
-        Error ->
-            Error
-    end.
-
-verify_new_start_matches_running(Name, Rate, Interval, NoOfProcesses, Group, State) ->
-    ExpectedNoOfProcesses = min(Rate, NoOfProcesses),
-    case {length(Group), State} of
-        {ExpectedNoOfProcesses, #{Name := #throttle_info{rate = Rate, interval = Interval}}} ->
-            {reply, {ok, already_started}, State};
-        {ExpectedNoOfProcesses, #{Name := #throttle_info{}}} ->
-            {reply, {error, wrong_reconfiguration}, State};
-        _ ->
-            {reply, {error, wrong_no_of_procs}, State}
-    end.
-
-run_cmd(Pid, stop) ->
-    amoc_throttle_process:stop(Pid);
-run_cmd(Pid, pause) ->
-    amoc_throttle_process:pause(Pid);
-run_cmd(Pid, resume) ->
-    amoc_throttle_process:resume(Pid).
+do_run_op(stop, Name, #throttle_info{pool_sup = PoolSup}, State) ->
```

Comment:
why don't we update process group here?

Reply:
This will stop the running processes and `pg` will get a monitor down message and remove them automatically.